helpful.el

```diff
@@ -920,6 +920,86 @@ POSITION-HEADS takes the form ((123 (defun foo)) (456 (defun bar)))."
   "Insert section break into helpful buffer."
   (insert "\n\n"))
 
+(defun helpful--calculate-references (source-path primitive-p)
+  "Calculate references for helpful--sym in SOURCE-PATH."
+  (when source-path
+    (let* ((buf (elisp-refs--contents-buffer source-path))
+           (positions
+            (if primitive-p
+                nil
+              (helpful--reference-positions
+               helpful--sym helpful--callable-p buf)))
+           (return-value (--map (helpful--outer-sexp buf it) positions)))
+      (kill-buffer buf)
+      return-value)))
+
+(defun helpful--make-manual-button (text)
+  "Make manual button with TEXT for helpful--sym."
+  (helpful--button
+   text
+   'helpful-manual-button
+   'symbol helpful--sym))
+
+(defun helpful--make-toggle-button (text buffer)
+  "Make toggle button with TEXT for helpful--sym in BUFFER."
+  (helpful--button
+   text
+   'helpful-toggle-button
+   'symbol helpful--sym
+   'buffer buffer))
+
+(defun helpful--make-set-button (text buffer)
+  "Make set button with TEXT for helpful--sym in BUFFER."
+  (helpful--button
+   text
+   'helpful-set-button
+   'symbol helpful--sym
+   'buffer buffer))
+
+(defun helpful--make-customize-button (text)
+  "Make customize button with TEXT for helpful--sym."
+  (helpful--button
+   text
+   'helpful-customize-button
+   'symbol helpful--sym))
+
+(defun helpful--make-references-button (text)
+  "Make references button with TEXT for helpful--sym."
+  (helpful--button
+   text
+   'helpful-all-references-button
+   'symbol helpful--sym
+   'callable-p helpful--callable-p))
+
+(defun helpful--make-edebug-button (text)
```

Comment:
The question of whether or not we should define functions for creating buttons is tricky. We did previously have them, and I factored them out in ade1aa5ec1f56ccb66d2fe994e49cb889dacf4ae.

I'm happy to have them if you think the code is clearer. However, they should be of the form `(defun helpful--make-foo-button (sym callable-p) ...)`. We're always calling most of these functions with the same value for `text`. `helpful--sym` and `helpful--callable-p` are both buffer-local variables, so these functions depend on the value of the current buffer, which could cause subtle bugs later.

Reply:
Happy to add the parameters, I only left them out because I thought that's just how Emacs code is written!

I do think the text should remain an argument, even though it will generally always be the same. Say you want to change the text -- where do you change it? Well, if it's an argument, then you can just look right in `helpful-update` and change it there. Other you have to search for it and find it tucked away in some crappy helper function.

So I propose the signature `(text sym callable-p)`.

Reply:
If you want to change the text, you can C-s to find the relevant string, and change it there :). If you want to change the text, and the button helper function is called in multiple places, you have to update all of them.

Pushing text logic into the button helper function also allows you to move additional checks out of `helpful-update`. For example, the edebug button:

```emacs-lisp
(helpful--button
          (if (helpful--edebug-p helpful--sym)
              "Disable edebug"
            "Enable edebug")
          'helpful-edebug-button
          'symbol helpful--sym)
```

If there's a function that gives me an edebug button, I don't want to duplicate the if statement every time I call it. I still think `(sym callable-p)` is the best fit here I'm afraid.